gossip/server.go

```diff
@@ -88,6 +88,10 @@ func (s *server) Gossip(argsI proto.Message) (proto.Message, error) {
 	if err != nil {
 		return nil, util.Errorf("local addr %s could not be converted to net.Addr: %s", args.LAddr, err)
 	}
+	// Verify that the client connection is valid and hasn't been closed.
+	if _, ok := s.lAddrMap[lAddr.String()]; !ok {
```

Comment:
We can't use the (distinguished) gossip server's listening address, because that's not known to the `net.Conn` object. The whole point of this ugly use of the sender's ephemeral local address is to find the one thing that both the gossip `Args` and the `net.Conn` can both have access to to match them up.

@bdarnell You're right about the client reconnects. I added a bool to the rpc context to disable that feature and I'm now having gossip always disable the RPC client cache, so it no longer tries to share RPC clients with the rest of the system. This way it can be sure about connects and disconnects and maintain a completely accurate set of incoming connections.